Assets/MixedRealityToolkit.Services/InputAnimation/InputRecordingService.cs

```diff
@@ -0,0 +1,317 @@
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Enable()
+        {
+            IsEnabled = true;
+            recordingBuffer = new InputAnimation();
+        }
+
+        /// <inheritdoc />
+        public override void Disable()
+        {
+            IsEnabled = false;
+            recordingBuffer = null;
+            ResetStartTime();
+        }
+
+        /// <inheritdoc />
+        public void StartRecording()
+        {
+            IsRecording = true;
+            if (UseBufferTimeLimit)
+            {
+                PruneBuffer();
+            }
+            if (!unlimitedRecordingStartTime.HasValue)
+            {
+                unlimitedRecordingStartTime = Time.time;
+            }
+        }
+
+        /// <inheritdoc />
+        public void StopRecording()
+        {
+            IsRecording = false;
+        }
+
+        /// <inheritdoc />
+        public override void LateUpdate()
+        {
+            if (IsEnabled)
+            {
+                if (IsRecording)
+                {
+                    if (UseBufferTimeLimit)
+                    {
+                        PruneBuffer();
+                    }
+
+                    RecordKeyframe();
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void DiscardRecordedInput()
+        {
+            if (IsEnabled)
+            {
+                recordingBuffer.Clear();
+                ResetStartTime();
+            }
+        }
+
+        /// <summary>
+        /// Record a keyframe at the given time for the main camera and tracked input devices.
+        /// </summary>
+        private void RecordKeyframe()
+        {
+            float time = Time.time;
+            var profile = InputRecordingProfile;
+
+            RecordInputHandData(Handedness.Left);
+            RecordInputHandData(Handedness.Right);
+            if (CameraCache.Main)
+            {
+                var cameraPose = new MixedRealityPose(CameraCache.Main.transform.position, CameraCache.Main.transform.rotation);
+                recordingBuffer.AddCameraPoseKey(time, cameraPose, profile.CameraPositionThreshold, profile.CameraRotationThreshold);
+            }
+        }
+
+        /// <summary>
+        /// Record a keyframe at the given time for a hand with the given handedness it is tracked.
+        /// </summary>
+        private bool RecordInputHandData(Handedness handedness)
+        {
+            float time = Time.time;
+            var profile = InputRecordingProfile;
+
+            var hand = HandJointUtils.FindHand(handedness);
+            if (hand == null)
+            {
+                recordingBuffer.AddHandStateKey(time, handedness, false, false);
+                return false;
+            }
+
+            bool isTracked = (hand.TrackingState == TrackingState.Tracked);
+
+            // Extract extra information from current interactions
+            bool isPinching = false;
```

Comment:
Do you know if it's actually possible to be pinching if not tracked?

Reply:
On the side of the input simulation this will never happen, because the hand device is destroyed when it stops tracking.

For the recorded input data itself, however, these are just separate animation tracks, there is no logic that prevents keyframes in the "IsPinching" curves when the "IsTracking" curve has a True value. It's not worth adding that kind of complication.

Nothing bad will happen when if you set the hand data to (isTracked=False, isPinching=True), the sim service just won't create a hand device and isPinching is ignored.